geo/geogfn: implement PostGIS compatible Area/Length/Perimeter

otan opened the pull request:
This commit implements Area, Length and Perimeter that is compatible
with PostGIS. There is a GEOMETRYCOLLECTION related bug with PostGIS,
where the Length of a Polygon includes the perimeter, and the Perimeter
of a LineString includes it's length, which seems like a bug. We've
copied it here for compatibility.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/48170)
<!-- Reviewable:end -->

sumeerbhola reviewed (approved):
> There is a GEOMETRYCOLLECTION related bug with PostGIS,
> where the Length of a Polygon includes the perimeter, and the Perimeter
> of a LineString includes it's length, which seems like a bug.

So the following comment is false, or they fixed POLYGON/MULTIPOLYGON but not a POLYGON contained in a GEOMETRYCOLLECTION?
https://postgis.net/docs/ST_Length.html Changed: 2.0.0 Breaking change -- in prior versions applying this to a MULTI/POLYGON of type geography would give you the perimeter of the POLYGON/MULTIPOLYGON. In 2.0.0 this was changed to return 0 to be in line with geometry behavior. Please use ST_Perimeter if you want the perimeter of a polygon

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 7 of 7 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48170#-:-M6fDCpUDWQwL-FU5BMV:b-6yh38l)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geographiclib/geographiclib.h, line 18 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48170#-M6fHC-u9l6Lkp2PAw9F:-M6fHC-u9l6Lkp2PAw9G:bzg37ap) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef8508f083aa9f1f0d20449b26d3f32231540871/pkg/geo/geographiclib/geographiclib.h#L18)):*
> ```objc
> 
> // CR_GEOGRAPHICLIB_InverseBatch computes the sum of the inverse geodesic problem
> // on an array of lat/lngs.
> ```

This comment is confusing. The inverse geodesic problem for two points is the length of the line connecting them and the azimuths. I am not sure what the "sum of the inverse geodesic problem" means. Also there is only one double result so presumably this is the length. Was this meant to be something like:
```
// ... computes the sum of the length of the lines represented by an array of lat/lngs.
void CR_GEOGRAPHICLIB_MultiLineLength(...
```
___
*[pkg/geo/geographiclib/geographiclib.go, line 68 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48170#-M6fIer56nvIXz1CZFRy:-M6fIer56nvIXz1CZFRz:b-ydo3gx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef8508f083aa9f1f0d20449b26d3f32231540871/pkg/geo/geographiclib/geographiclib.go#L68)):*
> ```Go
> // InverseBatch solves the inverse geodesic problem between all points provided.
> // This is intended for use for LineStrings. LinearRings/Polygons should use "AreaAndPerimeter".
> // Returns the sum of the s12 (distance in meters) units.
> ```

same comment about the code comment and the function name.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
Yep, the GEOMETRYCOLLECTION is still broken.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48170#-:-M6g69RoDNzB4iDsbtAE:b9khard)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @sumeerbhola)

___
*[pkg/geo/geographiclib/geographiclib.h, line 18 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48170#-M6fHC-u9l6Lkp2PAw9F:-M6g6QrJDfhEUh8eQIrz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef8508f083aa9f1f0d20449b26d3f32231540871/pkg/geo/geographiclib/geographiclib.h#L18)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This comment is confusing. The inverse geodesic problem for two points is the length of the line connecting them and the azimuths. I am not sure what the "sum of the inverse geodesic problem" means. Also there is only one double result so presumably this is the length. Was this meant to be something like:
```
// ... computes the sum of the length of the lines represented by an array of lat/lngs.
void CR_GEOGRAPHICLIB_MultiLineLength(...
```
</blockquote></details>

Done.
___
*[pkg/geo/geographiclib/geographiclib.go, line 68 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48170#-M6fIer56nvIXz1CZFRy:-M6g6iXe8zyR3YgEJFwR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef8508f083aa9f1f0d20449b26d3f32231540871/pkg/geo/geographiclib/geographiclib.go#L68)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

same comment about the code comment and the function name.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

otan committed:
geo/geogfn: implement PostGIS compatible Area/Length/Perimeter

This commit implements Area, Length and Perimeter that is compatible
with PostGIS. There is a GEOMETRYCOLLECTION related bug with PostGIS,
where the Length of a Polygon includes the perimeter, and the Perimeter
of a LineString includes it's length, which seems like a bug. We've
copied it here for compatibility.

Release note: None

otan commented:
bors r=sumeerbhola
